### PR TITLE
Apply solution-wide refactorings

### PR DESCRIPTION
**Pull request by tygyh**
Most of the refactorings are automated and can be found suggested in Jetbrains' IDEs. This PR will probably have conflicts with most other branches, so I'm in no hurry to have it merged.

### CODE REVIEW

**Comment by JosJuice** [comment]
This PR is way too large to review. You'll need to split it up if you want it to have any chance of being merged.

**Comment by tygyh** [comment]
> This PR is way too large to review. You'll need to split it up if you want it to have any chance of being merged.

You can look at the separate commits and pick out the parts you like. The commit messages hopefully give a good description of what has been done.

**Comment by BhaaLseN** [comment]
I did a quick scan over most of this, and I wonder if we even _want_ some of those.

First of all: yes, please make individual PRs for this, since the chances of getting this monster merged are close to zero.

Theres a few that I could see getting merged pretty easily, like the cast changes. Many of them are just left-overs that we didn't get to yet.
Maybe also the `constexpr` ones, which seem like low-hanging fruit in some cases.

I personally like structured bindings, but by the time we wrote much of the code that uses `std::tie(...)` now they weren't fully supported everywhere. It's only test code though, so it is of negligible consequence.
Same for the various upgrades that belong to newer C++ versions; we probably want those in there at some point.
Or the removal of unused includes (which likely affects compile times.)

Some of them seem like a big source of contention. I personally think that `Config::Get(...)` is more expressive than just `Get(...)` even when it would be in scope already and can be shortened.
Or keeping `if`/`else if`/`else` chains intact even though they terminate the method by `return` or similar.
Those likely benefit from codifying it in `.editorconfig` (if we can) or `.clang-format`. Or `Contributing.md` if there is no other way of making this tool-assisted in a semi-standardized way.

And I wonder if slapping `const` on methods that mutate state is correct (see `HttpRequest::Impl` for example, which _does_ mutate the internal curl options, but it doesn't change fields by setting them to some other value.)

The `const` parameter changes seem a bit weird, there's a few commits each and they don't seem to fully match their commit message (or look like they could just be combined into a single commit.)

Are those changes really fully tool assisted? Some of them look like you had to adjust things by hand (some of the structured binding changes, or the opposing methods for boolean parameters.)

Also, the fact alone that I had to write this comment in notepad then copy it into the review is a pretty good indicator (at least for me) that this PR is too large, just reinforcing what Jos already said.

**Comment by tygyh** [comment]
> Are those changes really fully tool assisted?

No, that's why I wrote 'Most'

**Comment by tygyh** [comment]
> Also, the fact alone that I had to write this comment in notepad then copy it into the review is a pretty good indicator (at least for me) that this PR is too large, just reinforcing what Jos already said.

I wrote this as a monster PR since I've been told when I make many small PRs I am DDOSing the CI system. I'll redo this with more digestable PRs.

**Comment by BhaaLseN** [comment]
> I wrote this as a monster PR since I've been told when I make many small PRs I am DDOSing the CI system. I'll redo this with more digestable PRs.

We appreciate your concern, but CI will manage ;)
Mainly because someone has to approve the CI run first, so it wont be too bad for us.

**Comment by tygyh** [comment]
> someone has to approve the CI run first

Is that a permanent thing, or is it until I've proven myself trustworthy? I'm just curious, not accusing you of gatekeeping. (Even though it is basically the definition of gatekeeping, but in a good way)

**Comment by BhaaLseN** [comment]
> > someone has to approve the CI run first
> 
> Is that a permanent thing, or is it until I've proven myself trustworthy? I'm just curious, not accusing you of gatekeeping. (Even though it is basically the definition of gatekeeping, but in a good way)

Just until we know you won't abuse it. This runs on our own infrastructure after all, and we don't want it to tank.

**Comment by mitaclaw** [comment]
The good:
- Remove unused includes: Yes, yes, yes a hundred times yes. This is the kind of thing automated tools are perfect at finding. Better compile times from this would be splendid. Still requires some manual review to make sure it's not violating the coding guideline for header grouping.
- Make method static: I can also get behind this change, though the automated approach could use some refinement, e.g. targeting random methods in the DSPInterpreter which just happen to not need the this pointer. Sometimes it's better to be consistent than correct, if that makes any sense.
- Use 'contains' method: I've made similar changes in the past on a smaller scale, so I can get behind this. It's unfortunate there's *so many* lines to change, however.
- Replace boolean constants with boolean literals: Well this is refreshingly small-scale. That one in SYSCONFSettings actually really could use the fix.
- Use structured bindings: I agree with BhaaLseN, this is a nice change.
- Convert 'reinterpret_cast' to 'static_cast': Yes. Will still require some manual review, but reinterpret_cast is a very sharp corner of C++.
- Replace redundant dereferencing: LOL, okay sure.

The bad:
- Use 'auto' style: Using `auto` everywhere possible is not wanted, see the coding guidelines: https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md#general-1.
- Replace bool parameters with opposing method names: This is an entirely subjective change that is unwarranted.
- Replace variables with structured bindings: This has the potential to make some code cleaner, but I don't trust such sweeping automated changes to code that works fine as-is.
- Iterate over values: Okay this is just `std::views` propaganda, and I love the ranges and views library! Totally unnecessary.
- Remove unnecessary template arguments: Similar to not always using 'auto', sometimes a vector or array having its template parameters written out explicitly helps reader comprehension. I would be interested in the `std::lock_guard`s having the `std::mutex` parameter removed, though.
- Remove redundant qualifiers: I'm pretty sure this one is straight up broken. It appears to be looking for any identifier in the current namespace named things like "Get" and assuming it can remove the namespace resolution preceding it if it finds one.

The ugly:
- Remove redundant 'else' keywords: I 100% agree that this is a superior coding style, and clang-tidy would agree with you as well, but this is needless. Code works fine as-is.
- Modernize range operations: As I said, I love the ranges library, but it's difficult to justify modifying so much working code just to make it prettier. I would give this a shot in a separate PR, but good luck.
- Remove redundant casts: One really nice thing about this commit is how many C-style casts it removes, but I just fear these casts may exist for a reason your automated tool missed. Maybe a compiler warning on a different platform? Maybe cross-platform issues with the WinAPI types? I just don't know about this one, since again it's working fine as-is.
- Convert to constant expressions: I like this a lot, but holy balls that is so many lines of change. It's infeasible to review this manually.
- Convert parameters to constants: Same issues as the last. Const-correctness is wonderful but you can't just expect this to be merged. You just can't.
- Replace functional-style casts with static casts: I really thought our coding guidelines say this is preferred, but I can't find it? Once again, insane line count. Doing this kind of thing hurts tons of open PRs, you know.
- Pass value by constant reference: I have sometimes in the past thought it would be a good idea to make function parameters be const, after all it is const correct to do so! Unfortunately, all it really does is make the function's prototype much longer, usually to the detriment of clang-format's line-width requirements. I can't get behind this change.
- Make methods constant: Again, I love const-correctness as much as the next guy, but come on. Way too huge of a commit.
- Make parameters constant: I'm having déjà-vu, wasn't this already done in a different commit? Wait, it just occurred to me that commit said it was supposed to be about const references? Well, same issue, this just makes clang-format look bad and is way too many functions changed at once.

Why in the world did you think a 47K lines changed automated refactoring would be a good idea for a PR, anyway? You have to understand this can't just be merged on a whim, even if all of the CI tests pass. It would be irresponsible.

**Comment by mitaclaw** [comment]
TLDR: The reason things are the way they are is because it's probably not a problem to leave them be and / or is subjective.

**Comment by mitaclaw** [comment]
Apparently `Config::Get` has been leaked into the global namespace, hence why your automated refactoring tool thinks it can remove the Config namespace in so many places. That's definitely an issue hiding somewhere in Dolphin.

**Comment by BhaaLseN** [comment]
I see a few `using namespace Common::something` sprinkled around, but they're mostly in the debugger/assembler-related sections (for `Common::GeckoAssembler`.) And I _hope_ they don't leak the whole `Common` namespace there.

**Comment by tygyh** [comment]
> Why in the world did you think a 47K lines changed automated refactoring would be a good idea for a PR, anyway?

I didn't think it had any chance to be merged as a whole. The conversation we're having right now was the goal.
I'm in the process of redoing each of the commits as separate PRs now. I'm doing the easy/desired ones first.

**Comment by Dentomologist** [comment]
I think Get(Info\<T\>) works without the Config prefix because of [ADL](https://en.cppreference.com/w/cpp/language/adl):

>Otherwise, for every argument in a function call expression its type is examined to determine the associated set of namespaces and classes that it will add to the lookup.
> 1) For arguments of fundamental type, the associated set of namespaces and classes is empty.
> 2) For arguments of class type (including union), the set consists of:
> ...
>   d) The innermost enclosing namespaces of the classes added to the set.

Info\<T\> is declared in the Config namespace, so that enables ADL to find the Get function. See the comment for the `endl(std::cout)` example on the linked page, which is the closest fit to our case here.

**Comment by mitaclaw** [comment]
That explains why I couldn't get other (empty parameter) functions from the config namespace to be recognized in the same way. Thanks.
